src/rust/ensogl/lib/core/src/display/scene/layer.rs

```diff
@@ -498,308 +498,435 @@ impl LayerModel {
-        if self.layers_depth_order_dirty.check() {
-            self.layers_depth_order_dirty.unset();
-            let model         = &mut *self.model.borrow_mut();
-            let layers_sorted = model.registry.iter().filter_map(|t|t.upgrade().map(|t|t.id));
-            let layers_sorted = layers_sorted.collect_vec();
-            let sorted_layers = model.layer_depth_order.unchecked_topo_sort(layers_sorted);
-            model.sorted_layers = sorted_layers;
-        }
-
-        if self.element_depth_order_dirty.check() {
-            self.element_depth_order_dirty.unset();
-            for layer in self.all() {
-                layer.update(&*self.global_element_depth_order.borrow())
+    fn remove_mask(&self) {
+        if let Some(mask) = &*self.mask.borrow() {
+            if let Some(mask) = mask.upgrade() {
+                mask.remove_parent(&self.children)
             }
         }
+        mem::take(&mut *self.mask.borrow_mut());
     }
 
-    /// Add depth-order dependency between two [`Layer`]s. Returns `true` if the dependency was
-    /// inserted successfully (was not already present), and `false` otherwise.
-    pub fn add_layers_order_dependency
-    (&self, below:impl Into<LayerId>, above:impl Into<LayerId>) -> bool {
-        let below = below.into();
-        let above = above.into();
-        let fresh = self.model.borrow_mut().layer_depth_order.insert_dependency(below,above);
-        if fresh { self.layers_depth_order_dirty.set(); }
-        fresh
+    /// Set all children layer of this layer. Old children layers will be unregistered.
+    pub fn set_children(&self, layers:&[&Layer]) {
+        self.remove_all_children();
+        for layer in layers {
+            self.add_child(layer)
+        }
     }
 
-    /// Remove depth-order dependency between two [`Layer`]s. Returns `true` if the dependency was
-    /// found, and `false` otherwise.
-    pub fn remove_layers_order_dependency
-    (&self, below:impl Into<LayerId>, above:impl Into<LayerId>) -> bool {
-        let below = below.into();
-        let above = above.into();
-        let found = self.model.borrow_mut().layer_depth_order.remove_dependency(below,above);
-        if found { self.layers_depth_order_dirty.set(); }
-        found
+    /// Set a mask layer of this layer. Old mask layer will be unregistered.
+    pub fn set_mask(&self, mask:&Layer) {
+        self.remove_mask();
+        *self.mask.borrow_mut() = Some(mask.downgrade());
+        mask.add_parent(&self.children);
     }
 
-    /// Add depth-order dependency between two [`LayerElement`]s in this layer. Returns `true`
+    /// Add depth-order dependency between two [`LayerItem`]s in this layer. Returns `true`
     /// if the dependency was inserted successfully (was not already present), and `false`
     /// otherwise.
-    pub fn add_elements_order_dependency
-    (&self, below:impl Into<LayerElement>, above:impl Into<LayerElement>) -> bool {
+    pub fn add_global_elements_order_dependency
```

Comment:
The docs does not explain difference between global and non-global order dependency.